client/lib/plugins/test/sanitize-section-content.js

```diff
@@ -0,0 +1,152 @@
+} );
+
+test( 'should bump up header levels', () => {
+	expect( clean( '<h1>👍</h1>' ) ).toBe( '<h3>👍</h3>' );
+	expect( clean( '<h2>👍</h2>' ) ).toBe( '<h3>👍</h3>' );
+	expect( clean( '<h3>👍</h3>' ) ).toBe( '<h3>👍</h3>' );
+	expect( clean( '<h4>👍</h4>' ) ).toBe( '<h4>👍</h4>' );
+} );
+
+test( 'should strip out <script> tags', () => expect( clean( '<script></script>' ) ).toBe( '' ) );
+
+/**
+ * The following tests have borrowed from the OWASP XSS Cheat Sheet
+ * @see https://www.owasp.org/index.php/XSS_Filter_Evasion_Cheat_Sheet
+ *
+ * Instead of pulling all of the tests the ones included are a sampling
+ * subjectively chosen to represent the visible attack vectors. For
+ * example, many of the attacks center around improper parsing of HTML
+ * which leads to unexpected `<script>` tags. However, we are using
+ * the browser's parser and not our own and thus don't need to test for
+ * some of those situations. We are already verifying that we strip
+ * <script> tags as well as non-whitelisted attributes, so in order for
+ * those security-related tests to fail our earlier tests would have
+ * also failed, and again we don't need to double-test them.
+ *
+ * Other tests were skipped because unless specifically verified as a
+ * viable URL in an `href` or `src` attribute, all other attribute
+ * data is encoded so that it should be impossible to end up with any
+ * javascript code. Therefore attacks through event handlers and the
+ * like should simply not be possible.
+ */
+test( 'should prevent known XSS attacks', () => {
+	expect( clean( "<IMG SRC=JaVaScRiPt:alert('XSS')>" ) ).toBe( '<img>' );
+
+	expect( clean( '<IMG SRC=javascript:alert(&quot;XSS&quot;)>' ) ).toBe( '<img>' );
+
+	expect( clean( '<IMG """><SCRIPT>alert("XSS")</SCRIPT>">' ) ).toBe( '<img>alert("XSS")"&gt;' );
+
+	expect(
+		clean(
+			// <IMG SRC=javascript:alert(
+			// 'XSS')>
+			'<IMG SRC=&#106;&#97;&#118;&#97;&#115;&#99;&#114;&#105;&#112;&#116;&#58;&#97;&#108;&#101;&#114;&#116;&#40;\n' +
+				'&#39;&#88;&#83;&#83;&#39;&#41;>'
+		)
+	).toBe( '<img>' );
+
+	expect( clean( "<img/onload=alert('XSS')>" ) ).toBe( '<img>' );
+
+	expect( clean( '¼script¾alert(¢XSS¢)¼/script¾' ) ).toBe( '¼script¾alert(¢XSS¢)¼/script¾' );
+
+	expect( clean( '<IFRAME SRC="javascript:alert(\'XSS\');"></IFRAME>' ) ).toBe( '' );
+
+	expect( clean( '<DIV STYLE="background-image: url(javascript:alert(\'XSS\'))">' ) ).toBe(
+		'<div></div>'
+	);
+
+	expect(
+		clean( '<!--[if gte IE 4]>\n' + " <SCRIPT>alert('XSS');</SCRIPT>\n" + ' <![endif]-->' )
+	).toBe( '' );
+
+	expect( clean( '<IMG SRC="javas<!-- -->cript:alert(\'XSS\')">' ) ).toBe( '<img>' );
+
+	expect(
+		clean( '<META HTTP-EQUIV="Set-Cookie" Content="USERID=<SCRIPT>alert(\'XSS\')</SCRIPT>">' )
+	).toBe( '' );
+
+	expect(
+		cleanNode(
+			'<A HREF="javascript:document.location=\'http://www.google.com/\'">XSS</A>'
+		).getAttribute( 'href' )
+	).toBeNull();
+} );
+
+test( 'should prevent backspace-based XSS attacks', () => {
```

Comment:
Is testing this stuff in `jsdom` sufficient? If we're worried about security, why should we assume the behavior of jsdom and whatever browser match (the spec or each other)?

This is probably an old discussion I'm ignorant of. Sorry :)

Reply:
would you recommend some integration test with Chrome? so far I think the behaviors are very standard and should not b implementation-specific, but I'm open to being more secure

Reply:
I'm not sure :)

On one hand, we're using the actual browser API for parsing, which means "no" (ha) inconsistencies between what we parse and what the browser does normally. (Except for browser bugs, mismatches between the user facing spec and internal implementations, etc.)

On the other, Calypso is now doing something different with HTML with respect to each browser. Before, in the process of sanitizing malicious HTML, we also normalized it to something simple that we could be reasonably confident all browsers would treat the same. We no longer normalize before sending it to the browser; we ask the browser to do that as part of its parsing (and our safelisting).

Worst case is that some browser is old enough to have buggy parsing but new enough to load a plugin page, and an onclick or something gets through. With the code in this PR, I don't know how that would be possible, but that's my concern.

This is all ameliorated by the fact that the plugins repo is semi-trusted, and all the HTML there gets run through KSES.

Reply:
Cool. I appreciate you raising this up and taking the time to critically evaluate it. I'm aware that it's "semi-trusted" but also don't want to rely too much on that! :smile: